File: Design-Patterns/Mediator/CSharp/Mediator/Employees/Employee.cs
using System;
using System.Collections.Generic;
using Mediator.Managers;

namespace Mediator.Employees
{
    public enum Mood { Thrilled, Good, Normal, Bad, Angry }

    public abstract class Employee
    {
        public string Name { get; protected set; }

        public int Health { get; protected set; }

        public Mood Mood { get; protected set; }

        public readonly List<Job> Jobs;

        protected Manager Manager;

        protected Employee(string name, Manager manager)
        {
            Name = name;
            Jobs = new List<Job>();
            Health = 100;
            Mood = Mood.Good;
            Manager = manager;
        }

        public virtual bool TakeJob(Job job, bool isForceMode = false)
        {
            if (Mood == Mood.Angry && !isForceMode) return false;

            Jobs.Add(job);

            if (Jobs.Count <= 3)
            {
                Mood = Mood.Thrilled;
            }
            else if (Jobs.Count <= 5)
            {
                Mood = Mood.Good;
            }
            else if (Jobs.Count <= 7)
            {
                Mood = Mood.Normal;
            }
            else if (Jobs.Count <= 9)
            {
                Mood = Mood.Bad;
            }
            else
            {
                Mood = Mood.Angry;
            }

            return true;
        }

        public virtual void DoJob()
        {
            if (Jobs.Count == 0) return;

            int index = new Random().Next(Jobs.Count);
            Job completedJob = Jobs[index];
            Jobs.RemoveAt(index);
            Manager.TransferJob(completedJob, this);
        }
    }
}

File: Design-Patterns/Mediator/Cpp/Mediator/employee.cpp
#include "employee.h"
#include <ctime>

employee::employee(std::string name, manager* manager)
{
	name_ = name;
	health_ = 100;
	mood_ = good;
	manager_ = manager;
}

employee::~employee()
{
	
}

std::string employee::get_name() const
{
	return name_;
}

int employee::get_health() const
{
	return health_;
}

moods employee::get_mood() const
{
	return mood_;
}

bool employee::take_job(job job, const bool is_force_mode)
{
	if (mood_ == angry && !is_force_mode) return false;

	jobs_.push_back(job);

	if (jobs_.size() <= 3)
	{
		mood_ = thrilled;
	}
	else if (jobs_.size() <= 5)
	{
		mood_ = good;
	}
	else if (jobs_.size() <= 7)
	{
		mood_ = normal;
	}
	else if (jobs_.size() <= 9)
	{
		mood_ = bad;
	}
	else
	{
		mood_ = angry;
	}

	return true;
}

void employee::do_job()
{
	if (jobs_.empty()) return;

	srand(time(nullptr));
	int index = rand() % jobs_.size();
	job completed_job = jobs_[index];
	jobs_.erase(jobs_.begin() + index);
	manager_->transfer_job(completed_job, this);
}
